### ISSUE TITLE
NS3002 false positive

### ISSUE THREAD

**Issue by AdamWillden** [nsubstitute/NSubstitute.Analyzers#118]
```cs
        private IMapper CreateMapperMock()
        {
            var mapper = Substitute.For<IMapper>();

            mapper
                .Map<InvitationDto>(Arg.Any<Invitation>())
                .Returns(_ =>
                {
                    //  NS3002 Can not find an argument of type ... to this call
                    _invitationDto.Roles = _.Arg<Invitation>().IdentityRoles; 

                    return _invitationDto;
                });
        }
```

**Comment by tpodolak** [comment]
Hi @AdamWillden thank you for reporting, could you show definition of IMapper interface?

**Comment by AdamWillden** [comment]
@tpodolak sorry... complete lack of info from me

It is AutoMapper: https://github.com/AutoMapper/AutoMapper/blob/master/src/AutoMapper/IMapper.cs

**Comment by AdamWillden** [comment]
P.S. thanks for creating this. Well worth the effort :-) and your hard work is appreciated.

**Comment by tpodolak** [comment]
This bug is caused due to the way we process and compare method arguments. In your case the method is defined as
````c#
public TResult Map<TResult>(object input);
````
When analyzer analyses usage of _.Arg<Invitation> it checks the method parameters of `Map` definition and it sees `object` instead of `Invitation` so it reports types incompatibility. 

What analyzer should do is to take parameter type from the invocation rather than from method symbol. In that case it will know that as you used Map with Invitation type and it is fine to do Arg<Invitation> in callback. This should be possible to fix but might be tricky to do correctly for named and optional parameters. I will try to look at this over the weekend

**Comment by AdamWillden** [comment]
Fantastic, thank you again
